File: regions/system/library/go/graphql-client/graphql_client.go
package graphqlclient

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
)

// ClientErrorKind はクライアントエラーの種別。
type ClientErrorKind int

const (
	// ClientErrorRequest はリクエストエラー。
	ClientErrorRequest ClientErrorKind = iota
	// ClientErrorDeserialization はデシリアライズエラー。
	ClientErrorDeserialization
	// ClientErrorGraphQl はGraphQLエラー。
	ClientErrorGraphQl
	// ClientErrorNotFound はリソース未検出エラー。
	ClientErrorNotFound
)

// String は ClientErrorKind の文字列表現を返す。
func (k ClientErrorKind) String() string {
	switch k {
	case ClientErrorRequest:
		return "request error"
	case ClientErrorDeserialization:
		return "deserialization error"
	case ClientErrorGraphQl:
		return "graphql error"
	case ClientErrorNotFound:
		return "not found"
	default:
		return "unknown error"
	}
}

// ClientError はGraphQLクライアントのエラー。
type ClientError struct {
	Kind    ClientErrorKind
	Message string
}

// Error は error インターフェースの実装。
func (e *ClientError) Error() string {
	return fmt.Sprintf("%s: %s", e.Kind, e.Message)
}

// NewRequestError は新しいリクエストエラーを生成する。
func NewRequestError(msg string) *ClientError {
	return &ClientError{Kind: ClientErrorRequest, Message: msg}
}

// NewDeserializationError は新しいデシリアライズエラーを生成する。
func NewDeserializationError(msg string) *ClientError {
	return &ClientError{Kind: ClientErrorDeserialization, Message: msg}
}

// NewGraphQlError は新しいGraphQLエラーを生成する。
func NewGraphQlError(msg string) *ClientError {
	return &ClientError{Kind: ClientErrorGraphQl, Message: msg}
}

// NewNotFoundError は新しいリソース未検出エラーを生成する。
func NewNotFoundError(msg string) *ClientError {
	return &ClientError{Kind: ClientErrorNotFound, Message: msg}
}

// GraphQlQuery はGraphQLクエリ。
type GraphQlQuery struct {
	Query         string         `json:"query"`
	Variables     map[string]any `json:"variables,omitempty"`
	OperationName string         `json:"operationName,omitempty"`
}

// GraphQlError はGraphQLエラー。
type GraphQlError struct {
	Message   string          `json:"message"`
	Locations []ErrorLocation `json:"locations,omitempty"`
	Path      []any           `json:"path,omitempty"`
}

// ErrorLocation はエラーの位置情報。
type ErrorLocation struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

// GraphQlResponse はGraphQLレスポンス。
type GraphQlResponse[T any] struct {
	Data   *T             `json:"data,omitempty"`
	Errors []GraphQlError `json:"errors,omitempty"`
}

// GraphQlClient はGraphQLクライアントのインターフェース。
type GraphQlClient interface {
	Execute(ctx context.Context, query GraphQlQuery, result any) (*GraphQlResponse[any], error)
	ExecuteMutation(ctx context.Context, mutation GraphQlQuery, result any) (*GraphQlResponse[any], error)
	Subscribe(ctx context.Context, subscription GraphQlQuery) (<-chan *GraphQlResponse[any], error)
}

// InMemoryGraphQlClient はメモリ内のGraphQLクライアント。
type InMemoryGraphQlClient struct {
	Responses     map[string]any
	subscriptions map[string][]any
	mu            sync.RWMutex
}

// NewInMemoryGraphQlClient は新しい InMemoryGraphQlClient を生成する。
func NewInMemoryGraphQlClient() *InMemoryGraphQlClient {
	return &InMemoryGraphQlClient{
		Responses:     make(map[string]any),
		subscriptions: make(map[string][]any),
	}
}

// SetResponse はオペレーション名に対するモック応答を設定する。
func (c *InMemoryGraphQlClient) SetResponse(operationName string, response any) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.Responses[operationName] = response
}

// SetSubscriptionEvents はオペレーション名に対するサブスクリプションイベント列を設定する。
func (c *InMemoryGraphQlClient) SetSubscriptionEvents(operationName string, events []any) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.subscriptions == nil {
		c.subscriptions = make(map[string][]any)
	}
	c.subscriptions[operationName] = events
}

// Execute はGraphQLクエリを実行する。
func (c *InMemoryGraphQlClient) Execute(_ context.Context, query GraphQlQuery, _ any) (*GraphQlResponse[any], error) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	resp, ok := c.Responses[query.OperationName]
	if !ok {
		return nil, fmt.Errorf("no response configured for operation: %s", query.OperationName)
	}
	return &GraphQlResponse[any]{Data: &resp}, nil
}

// ExecuteMutation はGraphQLミューテーションを実行する。
func (c *InMemoryGraphQlClient) ExecuteMutation(ctx context.Context, mutation GraphQlQuery, result any) (*GraphQlResponse[any], error) {
	return c.Execute(ctx, mutation, result)
}

// Subscribe はGraphQLサブスクリプションを実行し、イベントチャンネルを返す。
func (c *InMemoryGraphQlClient) Subscribe(_ context.Context, subscription GraphQlQuery) (<-chan *GraphQlResponse[any], error) {
	key := subscription.OperationName
	if key == "" {
		key = subscription.Query
	}
	c.mu.RLock()
	events, ok := c.subscriptions[key]
	c.mu.RUnlock()
	if !ok {
		events = nil
	}
	ch := make(chan *GraphQlResponse[any], len(events))
	for _, event := range events {
		e := event
		ch <- &GraphQlResponse[any]{Data: &e}
	}
	close(ch)
	return ch, nil
}

// GraphQlHttpClient はHTTPベースのGraphQLクライアント。
type GraphQlHttpClient struct {
	endpoint   string
	headers    map[string]string
	httpClient *http.Client
}

// NewGraphQlHttpClient は新しい GraphQlHttpClient を生成する。
func NewGraphQlHttpClient(endpoint string, headers map[string]string) *GraphQlHttpClient {
	h := headers
	if h == nil {
		h = make(map[string]string)
	}
	return &GraphQlHttpClient{
		endpoint:   endpoint,
		headers:    h,
		httpClient: &http.Client{},
	}
}

// Execute はGraphQLクエリをHTTPで実行する。
func (c *GraphQlHttpClient) Execute(ctx context.Context, query GraphQlQuery, _ any) (*GraphQlResponse[any], error) {
	return c.send(ctx, query)
}

// ExecuteMutation はGraphQLミューテーションをHTTPで実行する。
func (c *GraphQlHttpClient) ExecuteMutation(ctx context.Context, mutation GraphQlQuery, _ any) (*GraphQlResponse[any], error) {
	return c.send(ctx, mutation)
}

// Subscribe はGraphQLサブスクリプションを実行する。HTTPでは非対応のためエラーを返す。
func (c *GraphQlHttpClient) Subscribe(_ context.Context, _ GraphQlQuery) (<-chan *GraphQlResponse[any], error) {
	return nil, NewRequestError("GraphQlHttpClient does not support subscriptions over HTTP")
}

func (c *GraphQlHttpClient) send(ctx context.Context, query GraphQlQuery) (*GraphQlResponse[any], error) {
	body, err := json.Marshal(query)
	if err != nil {
		return nil, NewRequestError(fmt.Sprintf("failed to marshal query: %s", err))
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.endpoint, bytes.NewReader(body))
	if err != nil {
		return nil, NewRequestError(fmt.Sprintf("failed to create request: %s", err))
	}
	req.Header.Set("Content-Type", "application/json")
	for k, v := range c.headers {
		req.Header.Set(k, v)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, NewRequestError(err.Error())
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, NewRequestError(fmt.Sprintf("failed to read response body: %s", err))
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, NewRequestError(fmt.Sprintf("status %d: %s", resp.StatusCode, string(respBody)))
	}

	var gqlResp GraphQlResponse[any]
	if err := json.Unmarshal(respBody, &gqlResp); err != nil {
		return nil, NewDeserializationError(err.Error())
	}

	return &gqlResp, nil
}

File: regions/system/library/dart/graphql_client/lib/src/graphql_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'graphql_query.dart';

abstract class GraphQlClient {
  Future<GraphQlResponse<T>> execute<T>(
    GraphQlQuery query,
    T Function(Map<String, dynamic>) fromJson,
  );

  Future<GraphQlResponse<T>> executeMutation<T>(
    GraphQlQuery mutation,
    T Function(Map<String, dynamic>) fromJson,
  );

  Stream<GraphQlResponse<T>> subscribe<T>(
    GraphQlQuery subscription,
    T Function(Map<String, dynamic>) fromJson,
  );
}

class InMemoryGraphQlClient implements GraphQlClient {
  final Map<String, Map<String, dynamic>> _responses = {};
  final Map<String, List<Map<String, dynamic>>> _subscriptionEvents = {};

  void setResponse(String operationName, Map<String, dynamic> response) {
    _responses[operationName] = response;
  }

  void setSubscriptionEvents(
      String operationName, List<Map<String, dynamic>> events) {
    _subscriptionEvents[operationName] = events;
  }

  @override
  Future<GraphQlResponse<T>> execute<T>(
    GraphQlQuery query,
    T Function(Map<String, dynamic>) fromJson,
  ) async {
    return _resolve(query, fromJson);
  }

  @override
  Future<GraphQlResponse<T>> executeMutation<T>(
    GraphQlQuery mutation,
    T Function(Map<String, dynamic>) fromJson,
  ) async {
    return _resolve(mutation, fromJson);
  }

  @override
  Stream<GraphQlResponse<T>> subscribe<T>(
    GraphQlQuery subscription,
    T Function(Map<String, dynamic>) fromJson,
  ) {
    final key = subscription.operationName ?? subscription.query;
    final events = _subscriptionEvents[key] ?? [];
    return Stream.fromIterable(
      events.map((e) => GraphQlResponse(data: fromJson(e))),
    );
  }

  GraphQlResponse<T> _resolve<T>(
    GraphQlQuery query,
    T Function(Map<String, dynamic>) fromJson,
  ) {
    final key = query.operationName ?? query.query;
    final response = _responses[key];
    if (response == null) {
      return GraphQlResponse<T>(
        errors: [GraphQlError(message: 'No response configured for: $key')],
      );
    }

    final errorsRaw = response['errors'] as List<dynamic>?;
    final errors = errorsRaw
        ?.map((e) => GraphQlError(message: e['message'] as String))
        .toList();

    final dataRaw = response['data'] as Map<String, dynamic>?;
    final data = dataRaw != null ? fromJson(dataRaw) : null;

    return GraphQlResponse<T>(data: data, errors: errors);
  }
}

class GraphQlHttpClient implements GraphQlClient {
  final String _endpoint;
  final Map<String, String> _headers;
  final http.Client _httpClient;

  GraphQlHttpClient(
    this._endpoint, {
    Map<String, String>? headers,
    http.Client? httpClient,
  })  : _headers = {
          'Content-Type': 'application/json',
          ...?headers,
        },
        _httpClient = httpClient ?? http.Client();

  @override
  Future<GraphQlResponse<T>> execute<T>(
    GraphQlQuery query,
    T Function(Map<String, dynamic>) fromJson,
  ) {
    return _send(query, fromJson);
  }

  @override
  Future<GraphQlResponse<T>> executeMutation<T>(
    GraphQlQuery mutation,
    T Function(Map<String, dynamic>) fromJson,
  ) {
    return _send(mutation, fromJson);
  }

  @override
  Stream<GraphQlResponse<T>> subscribe<T>(
    GraphQlQuery subscription,
    T Function(Map<String, dynamic>) fromJson,
  ) {
    throw ClientError.request(
      'Subscriptions are not supported over HTTP',
    );
  }

  Future<GraphQlResponse<T>> _send<T>(
    GraphQlQuery query,
    T Function(Map<String, dynamic>) fromJson,
  ) async {
    final body = <String, dynamic>{
      'query': query.query,
    };
    if (query.variables != null) {
      body['variables'] = query.variables;
    }
    if (query.operationName != null) {
      body['operationName'] = query.operationName;
    }

    final http.Response response;
    try {
      response = await _httpClient.post(
        Uri.parse(_endpoint),
        headers: _headers,
        body: jsonEncode(body),
      );
    } on Exception catch (e) {
      throw ClientError.request('HTTP request failed: $e');
    }

    if (response.statusCode == 404) {
      throw ClientError.notFound('Resource not found: ${response.statusCode}');
    }

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw ClientError.request(
        'HTTP ${response.statusCode}: ${response.body}',
      );
    }

    final Map<String, dynamic> json;
    try {
      json = jsonDecode(response.body) as Map<String, dynamic>;
    } on FormatException catch (e) {
      throw ClientError.deserialization(
        'Failed to parse response JSON: $e',
      );
    }

    final errorsRaw = json['errors'] as List<dynamic>?;
    if (errorsRaw != null && errorsRaw.isNotEmpty) {
      final errors = errorsRaw.map((e) {
        final map = e as Map<String, dynamic>;
        final locationsRaw = map['locations'] as List<dynamic>?;
        final locations = locationsRaw?.map((l) {
          final loc = l as Map<String, dynamic>;
          return ErrorLocation(loc['line'] as int, loc['column'] as int);
        }).toList();
        final path = map['path'] as List<dynamic>?;
        return GraphQlError(
          message: map['message'] as String,
          locations: locations,
          path: path,
        );
      }).toList();

      final dataRaw = json['data'] as Map<String, dynamic>?;
      final data = dataRaw != null ? fromJson(dataRaw) : null;
      return GraphQlResponse<T>(data: data, errors: errors);
    }

    final dataRaw = json['data'] as Map<String, dynamic>?;
    if (dataRaw == null) {
      throw ClientError.deserialization('Response missing "data" field');
    }

    return GraphQlResponse<T>(data: fromJson(dataRaw));
  }
}
